cpmpy/expressions/globalfunctions.py

```diff
@@ -285,8 +287,19 @@ def decompose_comparison(self, cpm_op, cpm_rhs):
 
         """
         arr, idx = self.args
-        return [(idx == i).implies(eval_comparison(cpm_op, arr[i], cpm_rhs)) for i in range(len(arr))] + \
-               [idx >= 0, idx < len(arr)], []
+        cons = []
+        # the decomposition posts `idx=i -> arr[i] <CMP_OP> cpm_rhs`
+        # for every  `i` in the intersection of array indices and the bounds of `idx`
+        lb, ub = idx.get_bounds()
+        new_lb, new_ub = max(lb, 0), min(ub, len(arr))
+        for i in range(new_lb, new_ub):
+            cond = idx == i
+            if is_true_cst(cond):
+                return [eval_comparison(cpm_op, arr[i], cpm_rhs)], []
+            elif is_false_cst(cond):
```

Comment:
Not strictly necesarry, as we also check this in the `implies` function of `BoolVal`, but fine to keep.
Would prefer to add a small comment tho

Reply:
Ah true, I only checked implies of `BoolVar`, which only checks if the consequent is fixed. But then I'd rather go for simple code depending on the (good) implementation of implies, forgoing the early quit. So I'll remove these. I did have to add an utils.implies function which is also safe for expressions. I think this is very usefull to make more of the globals safe for constants. You can let me know what you think of the final decomposition